File: jvm/src/test/java/com/nawforce/apexparser/ApexVisitorTest.java
package com.nawforce.apexparser;

import org.antlr.v4.runtime.CommonTokenStream;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.StringReader;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ApexVisitorTest {

    static class TestVisitor extends ApexParserBaseVisitor<Integer> {
        public Integer methodCount = 0;

        @Override
        public Integer visitMethodDeclaration(ApexParser.MethodDeclarationContext ctx) {
            this.methodCount += 1;
            return 1 + super.visitChildren(ctx);
        }

        @Override
        public Integer defaultResult() {
            return 0;
        }
    }

    @Test
    void testVisitorIsVisited() throws IOException {
        ApexLexer lexer = new ApexLexer(new CaseInsensitiveInputStream(new StringReader(
                "public class Hello { public void func(){} }")));
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        ApexParser parser = new ApexParser(tokens);
        ApexParser.CompilationUnitContext context = parser.compilationUnit();

        TestVisitor visitor = new TestVisitor();
        visitor.visit(context);

        assertEquals(visitor.methodCount.intValue(), 1);
    }

}

File: npm/src/__tests__/ApexVisitorTest.ts
import { ApexLexer } from "../ApexLexer";
import { ApexParser, MethodDeclarationContext } from "../ApexParser";
import { CaseInsensitiveInputStream } from "../CaseInsensitiveInputStream"
import { CommonTokenStream } from 'antlr4ts';
import { AbstractParseTreeVisitor } from 'antlr4ts/tree/AbstractParseTreeVisitor';
import { ThrowingErrorListener } from "../ThrowingErrorListener";
import { ApexParserVisitor } from "../ApexParserVisitor";

class TestVisitor extends AbstractParseTreeVisitor<number> implements ApexParserVisitor<number> {
    public methodCount = 0

    visitMethodDeclaration(ctx: MethodDeclarationContext): number {
        this.methodCount += 1;
        return 1 + super.visitChildren(ctx)
    }

    defaultResult() {
        return 0
    }
}

test('Vistor is visited', () => {

    let lexer = new ApexLexer(new CaseInsensitiveInputStream("test.cls", "public class Hello { public void func(){} }"))
    let tokens = new CommonTokenStream(lexer);

    let parser = new ApexParser(tokens)

    parser.removeErrorListeners()
    parser.addErrorListener(new ThrowingErrorListener());

    const cu = parser.compilationUnit()
    const visitor = new TestVisitor()
    visitor.visit(cu)

    expect(visitor.methodCount).toBe(1)
})
